lotman/lotman_linux.go

```diff
@@ -314,9 +333,580 @@ func getFederationIssuer() (string, error) {
+	}
+
+	// policyLotMap will hold the lots defined in the configuration file (if any) provided by the cache admin
+	policyLotMap := make(map[string]Lot)
+	for _, lot := range policy.Lots {
+		policyLotMap[lot.LotName] = lot
+	}
+
+	var lotMap map[string]Lot
+	if discoverPrefixes {
+		directorLotMap, err := configLotsFromFedPrefixes(nsAds)
+		if err != nil {
+			return internalLots, errors.Wrap(err, "error configuring lots from federation prefixes")
+		}
+
+		// Handle potential need to merge discovered namespaces with provided configuration
+		if shouldMerge {
+			log.Debug("Merging lot configuration from discovered namespaces with configured lots")
+			lotMap, err = mergeLotMaps(directorLotMap, policyLotMap)
+			if err != nil {
+				return internalLots, errors.Wrap(err, "error merging discovered namespaces with configured lots")
+			}
+			log.Tracef("Merged lot configuration: %+v", lotMap)
+		} else {
+			lotMap = make(map[string]Lot)
+			// first set things up with the director lots, then overwrite with the policy lots.
+			// This allows cache admins to override any discovered lots with their own configuration.
+			for key, value := range directorLotMap {
+				lotMap[key] = value
+			}
+			for key, value := range policyLotMap {
+				lotMap[key] = value
+			}
+		}
+	} else {
+		lotMap = policyLotMap
+	}
+
+	cacheDisks := param.Cache_DataLocations.GetStringSlice()
+	log.Tracef("Cache data locations being tracked by Lotman: %v", cacheDisks)
+	var totalDiskSpaceB uint64
+	for _, disk := range cacheDisks {
+		diskSpace, _, err := getDiskUsage(disk)
+		if err != nil {
+			return internalLots, errors.Wrapf(err, "error getting disk usage for filesystem path %s", disk)
+		}
+		totalDiskSpaceB += diskSpace
+	}
+
+	// Now guarantee our special "default" and "root" lots if the user hasn't provided them
+	// in the config. For now, these lots must always exist because they're used to make sure
+	// all data is tied to a lot (default) and that the requirement of a root lot is satisfied
+	// without allowing discovered lots to gain rootly status.
+	federationIssuer, err := getFederationIssuer()
+	if err != nil {
+		return internalLots, errors.Wrap(err, "Unable to determine federation issuer which is needed by Lotman to determine lot ownership")
+	}
+	if federationIssuer == "" {
+		return internalLots, errors.New("The detected federation issuer, which is needed by Lotman to determine lot/namespace ownership, is empty")
+	}
+	rootDedGB := bytesToGigabytes(totalDiskSpaceB)
+	zero := float64(0)
+	if _, exists := lotMap["default"]; !exists {
+		lotMap["default"] = Lot{
+			LotName: "default",
+			Owner:   federationIssuer,
+			Parents: []string{"default"},
+			MPA: &MPA{
+				// Set default values to 0 and let potential reallocation happen later.
+				DedicatedGB:     &zero,
+				OpportunisticGB: &zero,
+				MaxNumObjects:   &Int64FromFloat{Value: 0}, // Purge plugin doesn't yet use this, set to 0.
```

Comment:
I'm a bit fuzzy on why I decided to make the Lot struct take all these pointers (it had something to do with the conversion of JSON in and out of Lotman), but the issue here is that I needed to define the float variable `zero` before I can reference it because Go doesn't let you reference literals. The other fields that take an int64 pointer can be instantiated directly, so they don't need the same setup.

That being said, I've already got the variable floating around there (ha) so I'm not opposed to using it if you'd prefer the consistency.

Reply:
Scratch that, there's a type error if I do that without yet another cast. I'd like to keep things the way they are for now if that's okay.